File: src/main/java/no/nav/familie/ks/oppslag/tilgangskontroll/TilgangsKontrollService.java
package no.nav.familie.ks.oppslag.tilgangskontroll;

import no.nav.familie.ks.oppslag.azure.AzureGraphService;
import no.nav.familie.ks.oppslag.azure.domene.Saksbehandler;
import no.nav.familie.ks.oppslag.egenansatt.EgenAnsattService;
import no.nav.familie.ks.oppslag.personopplysning.domene.Personinfo;
import no.nav.familie.ks.oppslag.tilgangskontroll.domene.Tilgang;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.stereotype.Service;

import static no.nav.familie.ks.oppslag.tilgangskontroll.domene.AdRoller.*;

@Service
public class TilgangsKontrollService {

    public static final String GEOGRAFISK = "GEOGRAFISK";
    public static final String TILGANGTILBRUKER = "tilgangtilbruker";
    public static final String TILGANGTILTJENESTEN = "tilgangtiltjenesten";
    public static final String TILGANGTILENHET = "tilgangtilenhet";

    private final AzureGraphService azureGraphService;
    private final EgenAnsattService egenAnsattService;

    private final static String ENHET = "ENHET";
    public final static String DISKRESJONSKODE_KODE6 = "SPSF";
    public final static String DISKRESJONSKODE_KODE7 = "SPFO";


    private static final Logger secureLogger = LoggerFactory.getLogger("secureLogger");
    private static final Logger LOG = LoggerFactory.getLogger(TilgangsKontrollService.class);


    @Autowired
    TilgangsKontrollService(
            AzureGraphService azureGraphService,
            EgenAnsattService egenAnsattService) {
        this.azureGraphService = azureGraphService;
        this.egenAnsattService = egenAnsattService;
    }

    @Cacheable(cacheNames = TILGANGTILBRUKER, key = "#saksbehandler.onPremisesSamAccountName.concat(#personFnr)", condition = "#personFnr != null && #saksbehandler.onPremisesSamAccountName != null")
    public Tilgang sjekkTilgang(String personFnr, Saksbehandler saksbehandler, Personinfo personInfo) {
        String NAVident = saksbehandler.getOnPremisesSamAccountName();
        String diskresjonskode = personInfo.getDiskresjonskode();

        if (DISKRESJONSKODE_KODE6.equals(diskresjonskode) && !harTilgangTilKode6(NAVident)) {
            secureLogger.info(NAVident + " har ikke tilgang til " + personFnr);
            return new Tilgang().withHarTilgang(false).withBegrunnelse(KODE6.name());
        } else if (DISKRESJONSKODE_KODE7.equals(diskresjonskode) && !harTilgangTilKode7(NAVident)) {
            secureLogger.info(NAVident + " har ikke tilgang til " + personFnr);
            return new Tilgang().withHarTilgang(false).withBegrunnelse(KODE7.name());
        }

        if (egenAnsattService.erEgenAnsatt(personFnr) && !harTilgangTilEgenAnsatt(NAVident)) {
            secureLogger.info(NAVident + " har ikke tilgang til egen ansatt " + personFnr);
            return new Tilgang().withHarTilgang(false).withBegrunnelse(EGEN_ANSATT.name());
        }

        return new Tilgang().withHarTilgang(true);
    }

    private boolean harTilgangTilKode7(String NAVident) {
        return false;
        //TODO hent roller fra token eller hent fra graph api til azuread
        //return ldapService.harTilgang(NAVident, KODE7.rolle);
    }

    private boolean harTilgangTilKode6(String NAVident) {
        return false;
        //TODO hent roller fra token eller hent fra graph api til azuread
        //return ldapService.harTilgang(NAVident, KODE6.rolle);
    }

    private boolean harTilgangTilEgenAnsatt(String NAVident) {
        return false;
        //TODO hent roller fra token eller hent fra graph api til azuread
        //return ldapService.harTilgang(NAVident, EGEN_ANSATT.rolle);
    }

}

File: src/main/java/no/nav/familie/integrasjoner/tilgangskontroll/TilgangsKontrollService.kt
package no.nav.familie.integrasjoner.tilgangskontroll

import no.nav.familie.integrasjoner.azure.domene.Saksbehandler
import no.nav.familie.integrasjoner.client.rest.AzureGraphRestClient
import no.nav.familie.integrasjoner.egenansatt.EgenAnsattService
import no.nav.familie.integrasjoner.personopplysning.domene.Personinfo
import no.nav.familie.integrasjoner.tilgangskontroll.domene.AdRoller
import no.nav.familie.integrasjoner.tilgangskontroll.domene.Tilgang
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.cache.annotation.Cacheable
import org.springframework.stereotype.Service

@Service
class TilgangsKontrollService @Autowired internal constructor(private val azureGraphRestClient: AzureGraphRestClient,
                                                              private val egenAnsattService: EgenAnsattService) {

    @Cacheable(cacheNames = [TILGANGTILBRUKER],
               key = "#saksbehandler.onPremisesSamAccountName.concat(#personFnr)",
               condition = "#personFnr != null && #saksbehandler.onPremisesSamAccountName != null")
    fun sjekkTilgang(personFnr: String?, saksbehandler: Saksbehandler, personInfo: Personinfo): Tilgang? {
        val NAVident = saksbehandler.onPremisesSamAccountName
        val diskresjonskode = personInfo.diskresjonskode
        if (DISKRESJONSKODE_KODE6 == diskresjonskode && !harTilgangTilKode6(NAVident)) {
            secureLogger.info("$NAVident har ikke tilgang til $personFnr")
            return Tilgang().withHarTilgang(false).withBegrunnelse(AdRoller.KODE6.name)
        }
        if (DISKRESJONSKODE_KODE7 == diskresjonskode && !harTilgangTilKode7(NAVident)) {
            secureLogger.info("$NAVident har ikke tilgang til $personFnr")
            return Tilgang().withHarTilgang(false).withBegrunnelse(AdRoller.KODE7.name)
        }
        if (egenAnsattService.erEgenAnsatt(personFnr) && !harTilgangTilEgenAnsatt(NAVident)) {
            secureLogger.info("$NAVident har ikke tilgang til egen ansatt $personFnr")
            return Tilgang().withHarTilgang(false).withBegrunnelse(AdRoller.EGEN_ANSATT.name)
        }
        return Tilgang().withHarTilgang(true)
    }

    private fun harTilgangTilKode7(NAVident: String?): Boolean {
        return false
        //TODO hent roller fra token eller hent fra graph api til azuread
//return ldapService.harTilgang(NAVident, KODE7.rolle);
    }

    private fun harTilgangTilKode6(NAVident: String?): Boolean {
        return false
        //TODO hent roller fra token eller hent fra graph api til azuread
//return ldapService.harTilgang(NAVident, KODE6.rolle);
    }

    private fun harTilgangTilEgenAnsatt(NAVident: String?): Boolean {
        return false
        //TODO hent roller fra token eller hent fra graph api til azuread
//return ldapService.harTilgang(NAVident, EGEN_ANSATT.rolle);
    }

    companion object {
        const val GEOGRAFISK = "GEOGRAFISK"
        const val TILGANGTILBRUKER = "tilgangtilbruker"
        const val TILGANGTILTJENESTEN = "tilgangtiltjenesten"
        const val TILGANGTILENHET = "tilgangtilenhet"
        private const val ENHET = "ENHET"
        const val DISKRESJONSKODE_KODE6 = "SPSF"
        const val DISKRESJONSKODE_KODE7 = "SPFO"
        private val secureLogger = LoggerFactory.getLogger("secureLogger")
        private val LOG = LoggerFactory.getLogger(TilgangsKontrollService::class.java)
    }

}
